File: dynamic_programming/279_perfect_squares.cc
#include <vector>

#include "limits.h"

// 2024/01/12: 自己实现

int numSquares(int n) {
  std::vector<int> dp(n + 1);
  for (int i = 1; i < n + 1; ++i) {
    int minn = INT_MAX;
    for (int j = 1; j * j <= i; ++j) { // 无需开根号, 直接比较
      minn = std::min(minn, dp[i - j * j]);
    }
    dp[i] = minn + 1;
  }
  return dp[n];
}

File: dynamic_programming/279_perfect_squares.py
import math


# 2026/06/05: 自己实现, debug 了一下.
def num_squares(n: int) -> int:
    dp = [float("inf")] * (n + 1)
    dp[0] = 0
    for i in range(1, n + 1):
        for j in range(1, int(math.sqrt(i)) + 1):
            dp[i] = min(dp[i - j * j] + 1, dp[i])
    return dp[-1]


print(num_squares(12))
print(num_squares(13))
